File: Array/leetcode/easy/LongestSubArrayWithKSum.go
package main

import "fmt"
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

// longestSubArrayWithKSum finds the length of the longest subarray with sum equal to k.
func longestSubArrayWithKSum(arr []int, k int) int {
    prefixSumMap := make(map[int]int)
    maxLength := 0
    sum := 0
    for i := 0; i < len(arr); i++ {
        sum += arr[i]
        value, exists := prefixSumMap[sum - k]
        if exists {
            currentLength := i - value
            maxLength = max(maxLength, currentLength)
        }

        if _, exists := prefixSumMap[sum]; !exists {
            prefixSumMap[sum] = i
        }

    }

    return maxLength
}

func longestSubArrayWithKSumMain() {
    arr := []int{1, 2, 3, -3, 1, 1, 1, 4, 2, -3}
    k := 5
    result := longestSubArrayWithKSum(arr, k)
    fmt.Printf("Length of the longest subarray with sum %d is: %d\n", k, result)
}

File: Array/leetcode/easy/LongestSubArrayWithKSum.py
def longest_subarray_with_k_sum(arr, k):
     maxLength = 0
     sum = 0
     prefixSumMap = {}
     
     
     for i in range(len(arr)):
         sum += arr[i]
         if (sum - k) in prefixSumMap:
             maxLength = max(maxLength, i - prefixSumMap[sum - k])
     
         if sum not in prefixSumMap:
             prefixSumMap[sum] = i
     
     return maxLength

# Example usage
if __name__ == "__main__":
    arr = [1, 2, 3, -3, 1, 1, 1, 4, 2, -3]
    k = 5
    result = longest_subarray_with_k_sum(arr, k)
    print(f"Length of the longest subarray with sum {k} is: {result}")